File: frameworks/core/interfaces/native/node/menu_item_modifier.cpp
/*
 * Copyright (c) 2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "core/interfaces/native/node/menu_item_modifier.h"

#include "base/log/log_wrapper.h"
#include "core/common/dynamic_module_helper.h"

namespace OHOS::Ace::NG {
namespace NodeModifier {
const ArkUIMenuItemModifier* GetMenuItemModifier()
{
    static const ArkUIMenuItemModifier* modifier = nullptr;
    if (modifier == nullptr) {
        auto* module = DynamicModuleHelper::GetInstance().GetDynamicModule("MenuItem");
        if (module == nullptr) {
            LOGF("Can't find MenuItem dynamic module");
            abort();
        }
        modifier = reinterpret_cast<const ArkUIMenuItemModifier*>(module->GetDynamicModifier());
    }
    return modifier;
}

const CJUIMenuItemModifier* GetCJUIMenuItemModifier()
{
    static const CJUIMenuItemModifier* modifier = nullptr;
    if (modifier == nullptr) {
        auto* module = DynamicModuleHelper::GetInstance().GetDynamicModule("MenuItem");
        if (module == nullptr) {
            LOGF("Can't find MenuItem dynamic module");
            abort();
        }
        modifier = reinterpret_cast<const CJUIMenuItemModifier*>(module->GetCjModifier());
    }
    return modifier;
}

#ifndef ACE_UNITTEST
const ArkUIMenuItemInnerModifier* GetMenuItemInnerModifier()
{
    static const ArkUIMenuItemInnerModifier* modifier = nullptr;
    if (modifier == nullptr) {
        auto* module = DynamicModuleHelper::GetInstance().GetDynamicModule("MenuItem");
        if (module == nullptr) {
            LOGF("Can't find MenuItem dynamic module");
            abort();
        }
        modifier =
            reinterpret_cast<const ArkUIMenuItemInnerModifier*>(module->GetCustomModifier("menuItemInnerModifier"));
    }
    return modifier;
}
#endif
} // namespace NodeModifier
} // namespace OHOS::Ace::NG

File: frameworks/bridge/declarative_frontend/ark_modifier/src/menu_item_modifier.ts
/*
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

class LazyArkMenuItemComponent extends ArkComponent {
  static module: MenuItemComponentModule | undefined = undefined;
  constructor(nativePtr: KNode, classType: ModifierType) {
   super(nativePtr, classType);
   if (LazyArkMenuItemComponent.module === undefined) {
     LazyArkMenuItemComponent.module = globalThis.requireNapi('arkui.components.arkmenuitem');
   }

   this.lazyComponent = LazyArkMenuItemComponent.module.createComponent(nativePtr, classType);
  }

  setMap(): void {
   this.lazyComponent._modifiersWithKeys = this._modifiersWithKeys;
  }

  selected(value: boolean): this {
   this.lazyComponent.selected(value);
   return this;
  }

  selectIcon(value: boolean | ResourceStr | SymbolGlyphModifier): this {
   this.lazyComponent.selectIcon(value);
   return this;
  }

  onChange(callback: (selected: boolean) => void): this {
   this.lazyComponent.onChange(callback);
   return this;
  }

  contentFont(value: Font): this {
   this.lazyComponent.contentFont(value);
   return this;
  }

  contentFontColor(value: ResourceColor): this {
   this.lazyComponent.contentFontColor(value);
   return this;
  }

  labelFont(value: Font): this {
   this.lazyComponent.labelFont(value);
   return this;
  }

  labelFontColor(value: ResourceColor): this {
   this.lazyComponent.labelFontColor(value);
   return this;
  }
}
class MenuItemModifier extends LazyArkMenuItemComponent implements AttributeModifier<MenuItemAttribute> {

  constructor(nativePtr: KNode, classType: ModifierType) {
    super(nativePtr, classType);
    this._modifiersWithKeys = new ModifierMap();
    this.setMap();
  }

  applyNormalAttribute(instance: MenuItemAttribute): void {
    ModifierUtils.applySetOnChange(this);
    ModifierUtils.applyAndMergeModifier<MenuItemAttribute, ArkMenuItemComponent, ArkComponent>(instance, this);
  }
}
